classification/class_task.py

```diff
@@ -88,10 +88,82 @@ def evaluate(self, test_loader):
                 total += self.batch_size
                 correct += (preds == y).sum().item()
         self.acc = correct / total
-        print(f'Validation accuracy: {self.acc}')
-        
+        print(f'{self.label} validation accuracy: {self.acc}')    
 
     def log_progress(self, step):
         write_weight_histograms(self.writer, self.net, step)
         write_loss_scalars(self.writer, self.loss_info, step)
         write_acc(self.writer, self.acc, step)
+
+
+class MLP_Classification():
+    def __init__(self, label, parameters):
+        super().__init__()
+        self.writer = SummaryWriter(comment=f"_{label}_training")
+        self.label = label
+        self.lr = parameters['lr']
+        self.hidden_units = parameters['hidden_units']
+        self.mode = parameters['mode']
+        self.batch_size = parameters['batch_size']
+        self.num_batches = parameters['num_batches']
+        self.x_shape = parameters['x_shape']
+        self.classes = parameters['classes']
+        self.save_model_path = f'{parameters["save_dir"]}/{label}_model.pt'
+        self.best_acc = 0.
+        self.dropout = parameters['dropout']
+        self.init_net(parameters)
+    
+    def init_net(self, parameters):
+        if not os.path.exists(parameters["save_dir"]):
+            os.makedirs(parameters["save_dir"])
+
+        model_params = {
+            'input_shape': self.x_shape,
+            'classes': self.classes,
+            'batch_size': self.batch_size,
+            'hidden_units': self.hidden_units,
+            'mode': self.mode,
+            'dropout': self.dropout
+        }
+        if self.dropout:
+            self.net = MLP_Dropout(model_params).to(DEVICE)
+        else:
+            self.net = MLP(model_params).to(DEVICE)
+        self.optimiser = torch.optim.Adam(self.net.parameters(), lr=self.lr)
+        self.scheduler = torch.optim.lr_scheduler.StepLR(self.optimiser, step_size=100, gamma=0.5)
+        print('MLP Parameters: ')
+        print(f'batch size: {self.batch_size}, input shape: {model_params["input_shape"]}, hidden units: {model_params["hidden_units"]}, output shape: {model_params["classes"]}, lr: {self.lr}')
+
+    def train_step(self, train_data):
+        self.net.train()
+        for _, (x, y) in enumerate(tqdm(train_data)):
+            x, y = x.to(DEVICE), y.to(DEVICE)
+            self.net.zero_grad()
+            self.loss_info = torch.nn.functional.cross_entropy(self.net(x), y, reduction='sum')
```

Comment:
@Eli-Persky, can you confirm whether sampling is required when performing prediction? If so, sample=True should be set in self.net(x).

Reply:
This is the MLP model - we only need sampling when predicting with the BNN, don't we?

Reply:
Sorry, meant to tag this line: https://github.com/tennisonliu/bayesian-neural-network/blob/cm_class_nns/classification/class_task.py#L72. which is sampling for BNN predictions.

Reply:
I guess so but I'm not really sure what that flag actually does tbh.

Reply:
ahh, so it would sample from the weight distributions, if the flag is not set, it will use the mean.

see this line: https://github.com/tennisonliu/bayesian-neural-network/blob/master/networks.py#L75

Reply:
Gotcha

Reply:
Sorry, I don't quite follow. Does sthg in the code need to change?

Reply:
ah yep, on this line https://github.com/tennisonliu/bayesian-neural-network/blob/cm_class_nns/classification/class_task.py#L72, it would be self.net(X, sample=True)